dolma-cccc/domain_counter.py

```diff
@@ -0,0 +1,239 @@
+from collections import defaultdict
+from concurrent.futures import ProcessPoolExecutor, as_completed, wait, FIRST_EXCEPTION
+from tqdm import tqdm
+import gc
+import signal
+import sys
+from datasets import Dataset, load_dataset
+
+
+def process_file(file_path):
+    suburl_counter = defaultdict(int)
+    text_length_counter = defaultdict(int)  # Additional counter based on text length
+
+    # First, count the total number of lines in the file for progress bar initialization
+    # with gzip.open(file_path, 'rt', encoding='utf-8') as f:
+    #     total_lines = sum(1 for _ in f)
+
+    # Re-open the file and process with progress bar
+    with gzip.open(file_path, 'rt', encoding='utf-8') as f:
+        with tqdm(desc=f"Processing {os.path.basename(file_path)}", leave=False,
+                  position=1) as line_progress:
+            for line in f:
+                try:
+                    data = json.loads(line)
+                    url = data.get('metadata', {}).get('warc_url', '')
+                    text = data.get('text', '')
+                    text_length = len(text)
+
+                    if url:
+                        suburls = extract_suburls(url)
+                        for suburl in suburls:
+                            suburl_counter[suburl] += 1
+                            text_length_counter[suburl] += text_length  # Increment based on text length
+
+                    line_progress.update(1)  # Update progress bar for each processed line
+                except json.JSONDecodeError:
+                    continue
+
+    gc.collect()  # Explicitly trigger garbage collection to free memory
+    return suburl_counter, text_length_counter
+
+
+def extract_suburls(url):
+    parsed_url = urlparse(url)
+    domain = parsed_url.netloc
+    path_parts = parsed_url.path.strip('/').split('/')
+
+    suburls = [domain, ]
+    current_path = domain
+
+    for part in path_parts[:-1]:
+        current_path = f"{current_path}/{part}"
+        suburls.append(current_path)
+
+    return suburls
+
+
+def merge_counters(counter_list):
+    merged_counter = defaultdict(int)
+    for counter in counter_list:
+        for suburl, count in counter.items():
+            merged_counter[suburl] += count
+    return merged_counter
+
+
+def process_all_files_in_parallel(root_folder):
+    all_files = []
+
+    for dirpath, _, filenames in os.walk(root_folder):
+        for filename in filenames:
+            if filename.endswith('.json.gz'):
+                all_files.append(os.path.join(dirpath, filename))
+
+    def signal_handler(sig, frame):
+        print("Interrupt received, terminating...")
+        executor.shutdown(wait=False)
+        sys.exit(1)
+
+    # Register the signal handler
+    signal.signal(signal.SIGINT, signal_handler)
+
+    with tqdm(total=len(all_files), desc="Total files", position=0) as file_progress:
+        with ProcessPoolExecutor(max_workers=40) as executor:  # Adjust max_workers based on your system
```

Comment:
This max worker should be a cli argument if you want people to adjust it themselves